jetty-proxy/src/test/java/org/eclipse/jetty/proxy/AsyncMiddleManServletTest.java

```diff
@@ -1551,6 +1556,82 @@ protected ContentTransformer newServerResponseContentTransformer(HttpServletRequ
         assertTrue(response.getHeaders().contains(PROXIED_HEADER));
     }
 
+    @Test
+    public void testDownstreamTransformationZipped() throws Exception
+    {
+        byte[] content1 = new byte[1024 * 1024];
+        ThreadLocalRandom.current().nextBytes(content1);
+        byte[] content2 = new byte[512 * 1024];
+        ThreadLocalRandom.current().nextBytes(content2);
+
+        startServer(new HttpServlet()
+        {
+            @Override
+            protected void service(HttpServletRequest request, HttpServletResponse response) throws IOException
+            {
+                response.setHeader(HttpHeader.CONTENT_ENCODING.asString(), "deflate");
+                ZipOutputStream zipOut = new ZipOutputStream(response.getOutputStream());
+                zipOut.putNextEntry(new ZipEntry("server-file1.rnd"));
+                zipOut.write(content1);
+                zipOut.closeEntry();
+                zipOut.putNextEntry(new ZipEntry("server-file2.rnd"));
+                zipOut.write(content2);
+                zipOut.closeEntry();
+                zipOut.finish();
+            }
+        });
+        startProxy(new AsyncMiddleManServlet()
+        {
+            @Override
+            protected ContentTransformer newServerResponseContentTransformer(HttpServletRequest clientRequest, HttpServletResponse proxyResponse, Response serverResponse)
+            {
+                return new AfterContentTransformer()
+                {
+                    @Override
+                    public boolean transform(Source source, Sink sink) throws IOException
+                    {
+                        // Un-zip and re-zip.
+
+                        ZipInputStream zipIn = new ZipInputStream(source.getInputStream());
+                        ZipOutputStream zipOut = new ZipOutputStream(sink.getOutputStream());
```

Comment:
Is this the wrong spot to do decompression and compression?  Shoudn't the client compress the data inbound and the server compress the outbound data normally ?